### ISSUE TITLE
UI notification changes issue

### ISSUE THREAD

**Issue by tomasfil** [reactivemarbles/DynamicData#544]
Hello, I ran into an issue propably with concurrency which I am unable to solve...

I have code that tracks statuses and current locations of palletes.

On top are statuses, which are fine and correct, but on bottom there are locations, which count is invalid only on UI side (WPF app)
![image](https://user-images.githubusercontent.com/34779382/147911822-7994d48e-c416-4059-91f2-7f2a7c092d78.png)

The counting is done by:
```
            publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .ObserveOnDispatcher()
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return machine.MachineDesc;
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return "EBS Nenalezeno";
                    }
                    return stockSearch.Get().Inventory;
                })
                .Group(g=>g)
                .SortBy(group=> group.Key)
                .Bind(out inventoryGroups)
                .Subscribe()
                .DisposeWith(d);
```

I know the data side is correct with this code:
```
            Observable.Interval(TimeSpan.FromSeconds(3))
                .Do(_ => Debug.WriteLine($"{jobHeader.DocName} {InventoryGroups.Sum(s=>s.Cache.Count)} - {TotalCount}"))
                .Subscribe()
                .DisposeWith(d);
```
output: `ASN-VBL294 34 - 34`

I did try to put Synchronize in between every method, but nothing helped. Only way this works correctly is by doing this:
```
            publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .ObserveOnDispatcher()
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return machine.MachineDesc;
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return "EBS Nenalezeno";
                    }
                    return stockSearch.Get().Inventory;
                })
                .Group(g=>g)
                .SortBy(group=> group.Key)
                .Bind(out inventoryGroups)
                .Subscribe()
                .DisposeWith(d);
```

But that is not desirable, I dont want IO operations running on Main thread.

Am I missing something?

I am on NET 5 and ReactiveUI 16.4.15.
I cannot transfer to NET 6 due to EF Core issue 27016.

**Comment by tomasfil** [comment]
I did try to move the location into class, so I skip possible issue with string not being reference type, but no luck there:
```
            publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return new JobLineWithActualLocation(s, machine.MachineDesc);
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return new JobLineWithActualLocation(s,"EBS Nenalezeno");
                    }
                    return new JobLineWithActualLocation(s, stockSearch.Get().Inventory);
                })
                .SortBy(s => s.ActualLocation)
                .Group(g=>g.ActualLocation)
                .ObserveOnDispatcher()
                .Bind(out inventoryGroups)
                .Subscribe()
                .DisposeWith(d);
```
![image](https://user-images.githubusercontent.com/34779382/147918052-eff1ba85-1385-43b0-b768-98b81256ca4a.png)

**Comment by tomasfil** [comment]
I am trying multiple iterations, but everything in the end breaks:
```
  publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return new JobLineWithActualLocation(s, machine.MachineDesc);
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return new JobLineWithActualLocation(s,"EBS Nenalezeno");
                    }
                    return new JobLineWithActualLocation(s, stockSearch.Get().Inventory);
                })
                .Group(g=>g.ActualLocation)
                .SortBy(s => s.Key)
                .ObserveOnDispatcher()
                .Synchronize()
                .Bind(out inventoryGroups)
                .Subscribe()
                .DisposeWith(d);
```

now I have 1 over the maximum instead of missing multiple... 
![image](https://user-images.githubusercontent.com/34779382/147918873-da468e04-1208-4a0e-99c1-739d2293a04e.png)

how is that possible?

**Comment by tomasfil** [comment]
I did try another approax with no success:

```
            publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return new JobLineWithActualLocation(s, machine.MachineDesc);
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return new JobLineWithActualLocation(s,"EBS Nenalezeno");
                    }
                    return new JobLineWithActualLocation(s, stockSearch.Get().Inventory);
                })
                .Bind(out jobLinesWithActualLocation)
                .Subscribe()
                .DisposeWith(d);

            jobLinesWithActualLocation.ToObservableChangeSet(x=>x.JobLine.Id)
                .Group(x=>x.ActualLocation)
                .ObserveOnDispatcher()
                .Bind(out inventoryGroups)
                .Subscribe()
                .DisposeWith(d);
```

Data is OK: 
`VBM049-DMR2103565 18 - 18 - 18`
from this watcher:
```
            Observable.Interval(TimeSpan.FromSeconds(3))
             .Do(_ => Debug.WriteLine($"{jobHeader.DocName} {InventoryGroups.Sum(s => s.Cache.Count)} - {TotalCount} - {jobLinesWithActualLocation.Count}"))
             .Subscribe()
             .DisposeWith(d);
```
but UI is not 
![image](https://user-images.githubusercontent.com/34779382/147920153-eb9bdd3f-f996-4c86-818a-9b9f52df8c09.png)

**Comment by tomasfil** [comment]
Another approach via AutoRefresh without success...
```
            publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return new JobLineWithActualLocation(s, machine.MachineDesc);
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return new JobLineWithActualLocation(s,"EBS Nenalezeno");
                    }
                    return new JobLineWithActualLocation(s, stockSearch.Get().Inventory);
                })
                .Bind(out jobLinesWithActualLocation)
                .Subscribe()
                .DisposeWith(d);

            jobLinesWithActualLocation.ToObservableChangeSet()
                .AutoRefresh(x=>x.ActualLocation)
                .GroupOn(x=>x.ActualLocation)
                .ObserveOnDispatcher()
                .Bind(out inventoryGroups)
                .Subscribe()
                .DisposeWith(d);
```

![image](https://user-images.githubusercontent.com/34779382/147921157-a36c07e3-858e-40f2-8336-b43192c2f737.png)

This is strange behaviour

**Comment by tomasfil** [comment]
Another test, same results once there are updates.
```
            publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return new JobLineWithActualLocation(s, machine.MachineDesc);
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return new JobLineWithActualLocation(s,"EBS Nenalezeno");
                    }
                    return new JobLineWithActualLocation(s, stockSearch.Get().Inventory);
                })
                .GroupOnPropertyWithImmutableState(x=>x.ActualLocation)
                .Synchronize()
                .ObserveOnDispatcher()
                .Bind(out inventoryGroups)
                .Subscribe()
                .DisposeWith(d);
```

Class for reference:
```
    public class JobLineWithActualLocation:ReactiveObject
    {
        public JobLineWithActualLocation(JobLine jobLine, string actualLocation)
        {
            JobLine = jobLine;
            ActualLocation = actualLocation;
        }

        [Reactive]
        public JobLine JobLine { get; set; }
        [Reactive]
        public string ActualLocation { get; set; }
    }
```

**Comment by tomasfil** [comment]
I managed to get the desired functionality like this:
```
            publishedJobLinesStoreCache
                .Filter(x => x is not null && JobLineQuery.IsValidFunc(x))
                .TransformAsync(async s =>
                {
                    using var repo = Locator.Current.GetService<IWmsRepository>();

                    if (s.Status==JobLineStatus.Picked && s.MachineId.HasValue)
                    {
                        var machine = await repo.MachineRepository.FindAsync(s.MachineId);
                        if(machine is not null && !string.IsNullOrEmpty( machine.MachineDesc))
                        {
                            return new JobLineWithActualLocation(s, machine.MachineDesc);
                        }
                    }

                    var jobService = new JobService(repo);
                    var stockSearch = await jobService.GetStockAsync(s);
                    if (stockSearch.Get() is null)
                    {
                        return new JobLineWithActualLocation(s,"EBS Nenalezeno");
                    }
                    return new JobLineWithActualLocation(s, stockSearch.Get().Inventory);
                })
                .ToCollection()
                .Select(col=>col.GroupBy(g=>g.ActualLocation).OrderBy(o=>o.Key))
                .ObserveOnDispatcher()
                .ToPropertyEx(this,x=>x.InventoryGroups)
                .DisposeWith(d);
```

But this way the ToCollection is huge resource waste because of the additional TransformAsync calls. Can someone explain to me, why the previous solutions dont work? They way I see it, they should. But it might be just lack of knowledge.

**Comment by RolandPheasant** [comment]
It is very hard for me to be able to understand what the desired behaviour is and what is happening. It would greatly help if you try to produce a breaking unit test, or if as you suggest the issue is currency, a minimal and simplified reproduction in a WPF demo app.

**Comment by tomasfil** [comment]
Ok I think I understand the issue. 
(I did cleanup this issue)

The Count of the final IGroup, or IGrouping doesn't raise INotifyPropertyChanged. Is there any way to achieve that?

I did create simple example where my issue is clearly visible:
https://github.com/tomasfil/ReactiveUI_Tests

The group gets created, deleted, but the count doesnt reflect changes. Any tip how to? .ToCollection() is expensive in my prod app, so I dont want to call it on every update (right now I have it setup like that)

**Comment by RolandPheasant** [comment]
The replication is a good example and has enabled me to diagnose.

The grouping is not designed for binding and provides no notify property change events. Also the group consists of an inner cache who's state changes as the group changes. This technically means there can is never a dynamic data update as any changes to a group result in a change to the inner cache rather than creating an updated new object.

The answer is to transform the group into an object which is suitable for binding, and within the transformed object listen to the cache count changed event and update a count .property.  I fixed it locally but cannot push to your branch as I do not have permissions.   

Instead I have created a gist https://gist.github.com/RolandPheasant/6a76877802bd5990b55a3bd766aa63d2

I think it should be self-explanatory but feel free to ask any questions.
